simexpal/util.py

```diff
@@ -110,33 +113,82 @@ def _validate_dict(dictionary, source, schema_path):
 							print("\nValidation errors in subschema [{}]:".format(cur_schema_index))
 						print(sub_error.message)
 					print()
-			sys.exit(1)
+			return False
+		return True
 
 	cur_file_path = os.path.abspath(os.path.dirname(__file__))
 
-	exp_yml_dict = read_setup_file(setup_file)
-	schema_path = os.path.join(cur_file_path, "schemes", "experiments.json")
-	_validate_dict(exp_yml_dict, "experiments.yml", schema_path)
+	validation_cache_dict = {}
+	try:
+		with open(os.path.join(basedir, 'validation.cache'), 'r') as f:
+			validation_cache_dict = json.load(f)
+	except FileNotFoundError:
+		pass
+
+	# Validate setup file and potentially cache results.
+	setup_file_path = os.path.join(basedir, setup_file)
+	setup_file_dict, setup_file_last_mod = read_setup_file(setup_file_path)
+	setup_file_schema_path = os.path.join(cur_file_path, "schemes", setup_file_schema)
 
+	if setup_file not in validation_cache_dict or setup_file_last_mod != validation_cache_dict[setup_file]:
+		setup_file_is_valid = _validate_dict(setup_file_dict, setup_file, setup_file_schema_path)
+
+		if setup_file_is_valid:
+			validation_cache_dict[setup_file] = setup_file_last_mod
+
+	# Validate launchers.yml file and potentially cache results.
 	try:
-		launchers_yml_dict = read_setup_file(os.path.expanduser('~/.simexpal/launchers.yml'))
-		schema_path = os.path.join(cur_file_path, "schemes", "launchers.json")
-		_validate_dict(launchers_yml_dict, "launchers.yml", schema_path)
+		launchers_yml_dict, launchers_yml_last_mod = read_setup_file(os.path.expanduser('~/.simexpal/launchers.yml'))
+		launchers_yml_schema_path = os.path.join(cur_file_path, "schemes", "launchers.json")
+
+		if 'launchers.yml' not in validation_cache_dict or launchers_yml_last_mod != validation_cache_dict['launchers.yml']:
+			launchers_yml_is_valid = _validate_dict(launchers_yml_dict, 'launchers.yml', launchers_yml_schema_path)
+
+			if launchers_yml_is_valid:
+				validation_cache_dict['launchers.yml'] = launchers_yml_last_mod
 	except FileNotFoundError:
 		pass
 
-	for exp in exp_yml_dict.get('experiments', []):
-		if exp.get('output', None) == 'stdout':
-			msg = "Specifying the stdout path via 'output: stdout' is deprecated and will be removed in future " \
-					"versions. Use 'stdout: out' instead."
-			warnings.warn(msg, DeprecationWarning)
+	writeback_cache = False
+	do_exit = False
+	try:
+		if setup_file_is_valid:
+			writeback_cache = True
+		else:
+			do_exit = True
+	except NameError:
+		pass
```

Comment:
This is not very elegant. This checks whether `setup_file_is_valid` is initialized? Why don't we set the correct default value at the top of the function (or use `None`)?

Reply:
I used ``None`` to avoid the ``try/except`` block